Add web section to Concurrency page (#4382)
dart-lang/site-www: src/_guides/language/concurrency/index.md

shares no mutable objects with other isolates.
To communicate with each other,
isolates use message passing.

> Many Dart apps use only one isolate (the _main isolate_),
> but you can create additional isolates,
> enabling parallel code execution on multiple processor cores.
> 

Reviewer: **suggestion:** Split the sentences. It still makes sense and reads better.
```suggestion
Many Dart apps use only one isolate, the _main isolate_.
You can create additional isolates. Using more isolates
enables parallel code execution on multiple processor cores.

```

Isolates are like threads or processes,
but each isolate has its own memory and a single thread running an event loop.

> {{site.alert.info}}
>   **Platform note:**
>     Isolates are implemented only on the [Dart Native platform][];
>     For information on the Dart Web platform,
>     see the [Concurrency on the web](#concurrency-on-the-web) section.
> {{site.alert.end}}

Reviewer: **suggestion:** Active voice.
```suggestion
{{site.alert.info}}
  **Platform note:**
    Only the [Dart Native platform][] implements Isolates.
    To learn more about the Dart Web platform,
    see the [Concurrency on the web](#concurrency-on-the-web) section.
{{site.alert.end}}
```

  Flutter doesn't support `Isolate.spawnUri()`.
{{site.alert.end}}

## Concurrency on the web

> All Dart apps can use `async-await`, `Future`, and `Stream`.

Reviewer 2: I am *slightly* worried about mixing "concurrency" and asynchronous computation, as if they are the same.
Asynchronous computations are still single-threaded (the opposite of concurrent), they just provide a practical way to yield control to other computations.

Maybe add:

> ... `Stream` for non-blocking, interleaved computations.
> Actual concurrency using isolates, however, is not ...

(Asynchronous functionality is great for *waiting for* other concurrent computations, because you don't block, but they are not themselves concurrent.)

Author: All the async=concurrency content in these PRs is because I misunderstood [this comment](https://github.com/dart-lang/site-www/issues/4042#issuecomment-1290683024) from the original issue, but I get it now! Close one 😅

> Isolates, however, are not available on the [Dart web platform][].

Reviewer: **suggestion:** Active voice / Agency.
```suggestion
[Dart web platform][] doesn't support Isolates.
```

Dart web apps can use [web workers][] to
run scripts in background threads
similar to isolates.
Web workers' functionality and capabilities
differ somewhat from isolates, though.

For instance, when web workers send data between threads,

> they copy the data back and forth.

Reviewer 2: (So do isolates, except for a the special case of `Isolate.exit`.)

Data copying can be very slow, though,
especially for large messages. 
So, rather than copying, isolates _transfer_ the memory
that holds the message in the exiting isolate
to the receiving isolate

> (specifically by `Isolate.exit()`).

Reviewer 2: Explaining isolate features doesn't feel like it fits here, under a heading of "Concurrency on the web". If it was compared to a similar web feature, then it might be relevant, but it

Drop everything that's just about isolates here, only focus on the web workers, or compare specific parts of web-workers to the similar isolate feature, but only if it is relevant in describing the web worker.

Reviewer: **suggestion:** This is a run-on sentence. Try to fix that and eliminate the parenthetical. The content of that parenthetical should stay, just not in parentheses.

Author: @lrhn I tried to [rewrite this](https://github.com/dart-lang/site-www/pull/4382/commits/d838272bf305a6ddc917d01ad8401de6d144ec0d#diff-07f3db7a96e07366164a1d378ebb716f2e0be3872e3060612bc74410a56ddd85R476-R487) to not go into so much detail on isolates, but still explain the comparison with web workers.
I skip mentioning `exit` and `run` outright and just said "APIs". Let me know if that doesn't make sense or is worse than before.

{% comment %}
Add ", and by extension `Isolate.run()`" to parenthesis after 2.19 and `run()` releases.
{% endcomment %}

How you create web workers and isolates also differs.
On the web, you can only create web workers by declaring
a separate program entrypoint and compiling that separately.
Natively, you can start isolates in a similar manner,
with separate sources, separate from existing isolates.

> More typically though, creation means branching from an existing isolate.

Reviewer 2: Maybe:

> You can only create web workers by declaring
> a separate program entrypoint and compiling that separately.
> Starting a web worker is similar to using `Isolate.spawnUri` to start an isolate,
> and is a more heavy-weight operation than using `Isolate.spawn`,
> which can reuse some of the same code and data as the spawning isolate
> (as mentioned in [Performance and isolate groups](#performance-and-isolate-groups)).

> How you create web workers and isolates also differs.
> On the web, you can only create web workers by declaring
> a separate program entrypoint and compiling that separately.
> Natively, you can start isolates in a similar manner,
> with separate sources, separate from existing isolates.
> More typically though, creation means branching from an existing isolate.

Reviewer: **suggestion:** Tighten and active voice. This builds on @lrhn 's revision.
```suggestion
The method to create web workers and isolates differs.
To create web workers, declare a separate program entry point
and compile that code separately.
A web worker starts a similar way to starting an isolate using `Isolate.spawnUri`.
This method uses more resources than using `Isolate.spawn`.
The latter method [can reuse some of the same code and data](#performance-and-isolate-groups) as the spawning isolate.
```

Author: > To create web workers, declare a separate program entry point
and compile that code separately.

"declare" is odd to me here. I feel like explaining a concept is decidedly not active, if that makes sense. We're not trying to literally get them to declare an entry point right now. I know always active is technically the rule in general but there has to be a caveat in this kind of situation?

Reviewer: **reply:** If you think it reads better as is. That's basically the caveat.

This makes both isolates part of the same isolate group,
and therefore much more lightweight 

> (as mentioned in [Performance and isolate groups](#performance-and-isolate-groups)).

Reviewer: **question:** @lrhn : Does these sentences need to be removed, then?

Reviewer 2: I'd say yes. It doesn't have any relevance to the web by itself, and it's not being compared or contrasted to any web feature.
